### ISSUE TITLE
Confusing issue (or possibly error) in HOW under "abusing beta-optimality"

### ISSUE THREAD

**Issue by HaraldKorneliussen** [Kindelia/HVM#87]
```
The same technique also applies for Inc. We start with the usual definition:

(Inc E)     = E
(Inc (O x)) = (I x)
(Inc (I x)) = (O (Inc x))
```

I don't understand this. It looks a lot like you're defining Inc 0 to be 0 in the first line. But surely if you increment 0 you get 1, right? Shouldn't it be

```
(Inc E) = (O E)
```

If so it looks a lot like defining numbers in a bijective base-2 scheme. While that is absolutely great and makes perfect sense to me, is it the usual definition? I thought that in lambda calculus they commonly define numbers with unary, i.e. just repeated application.

**Comment by VictorTaelin** [comment]
Yes, in the lambda calculus, they commonly define numbers in unary, mostly because that is interesting in theory, but a binary definition is still valid, and quite interesting, since it gives us an asymptotically efficient representation of big-ints using algebraic datatypes. Of course, such representation isn't as efficient as arrays of machine ints, but it is way more flexible, so it has its uses. For example, it features O(1) pattern-matching, is prone to fusion (as shown on that tutorial) can be used to encode [var-base](https://github.com/Kindelia/HVM/issues/82) integers, and is in general much simpler in code, which also makes proofs easier.

Answering your question, yes, `(Inc E) = (O E)` would make sense too. The reason I have `(Inc E) = E` is that this implementation is meant to wrap-around on overflow (i.e., `inc(15) = 0` when you have 3 bits). In your implementation, the behavior would be to grow on overflow (i.e., `inc(15) = 16`, increasing from 3 to 4 bits). Both are valid and useful on different scenarios.
